HW1/P4/P4.py

```diff
@@ -0,0 +1,75 @@
+import time
+import findspark
+findspark.init('/home/zelong/spark')
+import pyspark
+from P4_bfs import bfs
+starttime = time.time()
+
+sc = pyspark.SparkContext('local','bfs')
+# Construct the graph from data
+data = sc.textFile('source.csv')
+pair = data.map(lambda x: x[1:-1].split("\",\""))
+movie_name_pair = pair.map(lambda x: (x[1],x[0]))
+group_by_movie = movie_name_pair.groupByKey().map(lambda x: (x[0],list(set(x[1]))))
+
+def assign_neighbor(x):
+    # x is an iterable contains all characters in the movie
+    result = [] 
+    for i in range(len(x)):
+	key = x[i]
+	value = x[0:i] + x[i+1:]
+	result.append( (key, value) )
+    return result 
+
+neighbor = group_by_movie.flatMap(lambda x: assign_neighbor(x[1]))
+graph = neighbor.groupByKey().map(lambda x: (x[0], list(set([i for j in x[1] for i in j]))))
+graph_dist = graph.map(lambda x: (x[0], ( 999999 , x[1]   )))
+
+
+#generate txt file
+f = open('P4.txt','w')
+
+#write result of "CAPTAIN AMERICA"
+output = bfs(graph_dist, 'CAPTAIN AMERICA', 6, sc )
+result = output[0].cache()
+num_iter = output[1] - 1
```

Comment:
I like how you not only kept track of total nodes connected to Captain America, but that you also kept track of the number of nodes by specific distance (0, 1, 2).